### PR TITLE
Export document layout

### PR DESCRIPTION
**Pull request by Jestemrique**
Class to export a specific document layout to PDF.

### CODE REVIEW

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			System.out.println("Files saved to: " + pathToSavePDFFile);
+		} catch (WebObjectsException e) {
+			e.printStackTrace();
+		}
+		
+	}
+	
+	
+	public static byte[] exportDocumentLayoutToPDF( String documentID, WebIServerSession session, String layoutName ) throws WebObjectsException {
+		System.out.println("Exporting layout.");
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Can you run the indent function in Eclipse? Some of it isn't aligned

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		
+		RWBean rwb = (RWBean)BeanFactory.getInstance().newBean("RWBean");
+
+	    // ID of the object we wish to execute
+	    rwb.setObjectID(documentID);
+	    
+	    // session used for execution
+	    rwb.setSessionInfo(session);
+	    
+	    RWInstance rwInstance = null;
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Unnecessary line-spacing here, group similar calls together - makes it more readable imo

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	
+	    RWInstance rwInstanceToExport = null;
+	    
+	    String layoutToExportKey = null;
+		try {
+			
+			rwInstance = rwb.getRWInstance();
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Indent

File: stand-alone/com/microstrategy/samples/users/CreateUser.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
+package com.microstrategy.samples.users;
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/users/CreateUser.java]: Looks like CreateUser found its way in here too! It's a sneaky one!

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 1 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done!

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 2 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done!

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    RWInstance rwInstance = null;
+    RWInstance rwInstanceToExport = null;
+    String layoutToExportKey = null;
+	try {
+		
+		rwInstance = rwb.getRWInstance();
+
+		RWDefinition rwDefinition = rwInstance.getDefinition();
+		
+		System.out.println("Document Name: " + rwInstance.getDefinition().getName());
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: These system.out dont seem necessary.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	String layoutKey = null;
+	String iterationName = null;
+	int layoutIteration = 0;
+	
+	do {
+		iterationName = rwDefinition.getLayout(layoutIteration).getName();
+		layoutKey = rwDefinition.getLayout(layoutIteration).getKey();
+		layoutIteration++;
+	}
+	while( !(layoutName.equals(iterationName)));
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: you need a circuit breaker here.

File: stand-alone/com/microstrategy/samples/users/CreateUser.java (hunk 4 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/users/CreateUser.java]: Done!

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  }
+	
+	
+  public static byte[] exportDocumentLayoutToPDF( String documentID, WebIServerSession session, String layoutName ) throws WebObjectsException {
+	System.out.println("Exporting layout.");
+	
+	RWBean rwb = (RWBean)BeanFactory.getInstance().newBean("RWBean");
+
+    // ID of the object we wish to execute
+    rwb.setObjectID(documentID);
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: indentation is inconsistent.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+		FileHelper.saveByteArrayToFile(exportLayoutResults, pathToSavePDFFile);
+		
+		System.out.println("Files saved to: " + pathToSavePDFFile);
+	} catch (WebObjectsException e) {
+		e.printStackTrace();
+	}
+  }
+	
+	
+  public static byte[] exportDocumentLayoutToPDF( String documentID, WebIServerSession session, String layoutName ) throws WebObjectsException {
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: You do not need a space between parenthasis and String; 'PDF( Str' -> 'PDF(Str'

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    String microstrategyUsername = "Administrator";
+    String microstrategyPassword = "";
+    
+    // Create our I-Server Session
+    WebIServerSession session = SessionManager.getSessionWithDetails(intelligenceServerName, projectName, microstrategyUsername, microstrategyPassword);
+    
+    //Document ID
+    String documentID = "C0AEF3624E8FE0A8E5CBC09ACBF51FA2";
+    
+    String pathToSavePDFFile = "/Users/mpastrana/trabajos/dev/eprcode/output-trash/layout_exported.pdf";
```
Comment (Dholmblad) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Remove hardcoded path as this is only relevant to developers environment.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 3 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 5 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: done

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 7 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 9 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: This path code here is on purpose. User can get the path to save the resulting file by any other means (parameter, etc.)
I've put a comment in the code. Tell me if I should put it in a specific way.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 8 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 6 above)
Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Need help here. What do you mean by a circuit breaker? Do I need to take into account endless loops and things like that?

Comment (Dholmblad) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: @Jestemrique yes you might never find a match, especially considering you are doing case-sensitive matching which may be unreliable.

Comment (tiagosiebler) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Hey @Jestemrique, perhaps a for loop instead, or set a max iteration limit before you `break;`. If a layout combo can't be found here, this loop would max out forever.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
+package com.microstrategy.samples.beans.rwbean;
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Please also clean up the file indent

Comment (Jestemrique) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: Done. Added the following to end while loop.
while( !(layoutName.equals(iterationName)) && layoutIteration < numberOfLayouts);

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	}
+  }
+	
+	
+  public static byte[] exportDocumentLayoutToPDF(String documentID, WebIServerSession session, String layoutName ) throws WebObjectsException {
+	System.out.println("Exporting layout.");
+	
+	RWBean rwb = (RWBean)BeanFactory.getInstance().newBean("RWBean");
+
+    // ID of the object we wish to execute
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: indent also needs a fix in this method

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+		RWDefinition rwDefinition = rwInstance.getDefinition();
+
+		//Get layout key from layout name.
+		layoutToExportKey = getLayoutKeyFromName(rwDefinition, layoutName);
+		
+		RWManipulation rwManipulation = rwInstance.getRWManipulator(true);
+		rwInstanceToExport = rwManipulation.setCurrentLayout(layoutToExportKey);
+	    
+		// set execution mode to PDF export
+	    rwb.setExecutionMode(EnumRWExecutionModes.RW_MODE_PDF);
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java]: indent is also inconsistent here

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 13]
```diff
@@ -18,7 +18,7 @@
 
   public static void main(String[] args) {
 	// Connectivity for the intelligence server
```
Comment (ddechent) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:20]: Comment is indented, should fall in line

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 14]
```diff
@@ -27,7 +27,7 @@ public static void main(String[] args) {
     WebIServerSession session = SessionManager.getSessionWithDetails(intelligenceServerName, projectName, microstrategyUsername, microstrategyPassword);
     
     //Document ID
-    String documentID = "C0AEF3624E8FE0A8E5CBC09ACBF51FA2";
+    String documentID = "6FB9F99C4FCF116AC88C8F927F15D2CC";
     
     //Path to save pdf file needs to be provided via "pathToSavePDFFile" var.
```
Comment (ddechent) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:32]: Should have one space between // and Path

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 15]
```diff
@@ -27,7 +27,7 @@ public static void main(String[] args) {
     WebIServerSession session = SessionManager.getSessionWithDetails(intelligenceServerName, projectName, microstrategyUsername, microstrategyPassword);
     
     //Document ID
```
Comment (ddechent) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:29]: Should have one space between // and Document ID

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 16]
```diff
@@ -18,7 +18,7 @@
 
   public static void main(String[] args) {
 	// Connectivity for the intelligence server
-    String intelligenceServerName = "10.23.5.242";
+    String intelligenceServerName = "10.23.5.32";
     String projectName = "MicroStrategy Tutorial";
```
Comment (ddechent) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:22]: Optional suggestion:
An alternate way to document hard-coded values is to use final String MY_STRING_NAME = "Name"; so line 22 would become:
final String PROJECT_NAME = "MicroStrategy Tutorial";

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+	    
+	    RWExportSettings exportSettings = rwInstance.getExportSettings();
+	    exportSettings.setMode(EnumRWExportModes.RW_EXPORT_CURRENT_LAYOUT);
+	    
+	    // Switch the current layout.
+	    rwInstanceToExport.setCurrentLayoutKey(layoutToExportKey);
+	    rwInstanceToExport.setAsync(false);
+	    rwInstanceToExport.pollStatus();
+		
+		} catch (WebBeanException | WebObjectsException e) {
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:85]: Indent is off here

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+import com.microstrategy.web.objects.rw.RWDefinition;
+import com.microstrategy.web.objects.rw.RWExportSettings;
+import com.microstrategy.web.objects.rw.RWInstance;
+import com.microstrategy.web.objects.rw.RWManipulation;
+
+public class ExportDocumentLayout {
+	static final String ISERVERNAME = "10.23.1.124";
+	static final String PROJECT_NAME = "MicroStrategy Tutorial";
+	static final String USER_NAME = "Administrator";
+	static final String PASSWORD = "";
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:21]: Indent is off here

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    // Document ID
+    String documentID = "769B8AB6471B182B44CB7089ECC2F607";
+    
+    // Path to save pdf file needs to be provided via "pathToSavePDFFile" var.
+    String pathToSavePDFFile = "/Users/mpastrana/trabajos/testlayout.pdf";
+    
+    // Layout to export.
+    String layoutName = "Layout 4";
+    
+	try {
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:36]: Also inconsistent indent. Are you using Eclipse's automatic indenting? If not, I recommend using that to clean this up in one action.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+	    rwInstanceToExport.setCurrentLayoutKey(layoutToExportKey);
+	    rwInstanceToExport.setAsync(false);
+	    rwInstanceToExport.pollStatus();
+		
+		} catch (WebBeanException | WebObjectsException e) {
+		e.printStackTrace();
+	}
+	
+    
+    // return rwInstance.getExportData();
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:90]: Dead comment, I'd remove that

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    return rwInstanceToExport.getExportData();
+
+  }
+
+
+  public static String getLayoutKeyFromName(RWDefinition rwDefinition, String layoutName) {
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:96]: Recommend short comment to describe purpose of this function relative to the rest of the document workflow.

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java (hunk 20 above)
Comment (tiagosiebler) [reply stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:89]: dead comment still there

File: stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+      exportSettings.setMode(EnumRWExportModes.RW_EXPORT_CURRENT_LAYOUT);
+
+      // Switch the current layout.
+      rwInstanceToExport.setCurrentLayoutKey(layoutToExportKey);
+      rwInstanceToExport.setAsync(false);
+      rwInstanceToExport.pollStatus();
+	
+	} catch (WebBeanException | WebObjectsException e) {
+	  e.printStackTrace();
+	}
```
Comment (tiagosiebler) [inline stand-alone/com/microstrategy/samples/beans/rwbean/ExportDocumentLayout.java:87]: indent of catch block

**Pull request merged**